README.md

```diff
@@ -735,16 +804,27 @@ The most flexible webhook GitHub offers is the `workflow_job` webhook, it includ
 This webhook should cover most people's needs, please experiment with this webhook first before considering the others.
 
 ```yaml
+apiVersion: actions.summerwind.dev/v1alpha1
 kind: RunnerDeployment
 metadata:
-   name: example-runners
+  name: example-runners
+  namespace: actions-runner-system
```

Comment:
```suggestion
```

I believe it's a good practice to separate the controller and runner namespace for security.
Just keep omitting it so that it would default to the `default` namespace

Reply:
Uhm, I decided to test whether this really was unnecessary, but when I removed it I started having some problems. It's late now, so I might have made some mistakes, I will continue this investigation tomorrow.

Reply:
@mumoshu I am not sure how to interpret the results of my experiment, but I tried removing the namespace from the `RunnerDeployment` and the `HorizontalRunnerAutoscaler` and it actually stopped working.

Just so you can understand the steps I took, I used this command to install arc:

```bash
helm upgrade --install --namespace actions-runner-system --create-namespace \
             --wait actions-runner-controller actions-runner-controller/actions-runner-controller \
             --set "githubWebhookServer.enabled=true,githubWebhookServer.ports[0].nodePort=33080"
```

Then I created these two resources below, originally with the namespace, then I removed them then I added them again.

```diff
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: example-runners
-  namespace: actions-runner-system
spec:
  template:
    spec:
      repository: vingarcia/ksql
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: example-runners
-  namespace: actions-runner-system
spec:
  scaleDownDelaySecondsAfterScaleOut: 300
  minReplicas: 1
  maxReplicas: 10
  scaleTargetRef:
    name: example-runners
  scaleUpTriggers:
  - githubEvent: {}
    duration: "30m"
```

After removing by running the delete kubectl command on both resources I was left with no RunnerDeployments and no HRAs, then I created them again without the namespace and provoked some workflowJob events, and it did not autoscale as it was autoscaling before. I checked and both resources were created in the `default` namespace then I opened the logs for the `actions-runner-controller` and the `actions-runner-controller-github-webhook-server` and provoked another workflowJob event (by pushing some changes to my repo). On the controller I got nothing out of the usual, but I got several errors on the webhook-server:

```logs
2022-05-17T23:21:06Z    ERROR   controllers.webhookbasedautoscaler      finding repository-wide runner  {"event": "workflow_job", "hookID": "353945701", "delivery": "069f0440-d638-11ec-8ec3-d741f7b474a4", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted"], "repository.name": "ksql", "repository.owner.login": "VinGarcia", "repository.owner.type": "User", "enterprise.slug": "", "action": "queued", "repository": "VinGarcia/ksql", "error": "cache contained <nil>, which is not an Object"}
github.com/actions-runner-controller/actions-runner-controller/controllers.(*HorizontalRunnerAutoscalerGitHubWebhook).getJobScaleUpTargetForRepoOrOrg
        /workspace/controllers/horizontal_runner_autoscaler_webhook.go:456
github.com/actions-runner-controller/actions-runner-controller/controllers.(*HorizontalRunnerAutoscalerGitHubWebhook).Handle
        /workspace/controllers/horizontal_runner_autoscaler_webhook.go:236
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2047
net/http.(*ServeMux).ServeHTTP
        /usr/local/go/src/net/http/server.go:2425
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2879
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1930
2022-05-17T23:21:06Z    ERROR   controllers.webhookbasedautoscaler      handling check_run event        {"event": "workflow_job", "hookID": "353945701", "delivery": "069f0440-d638-11ec-8ec3-d741f7b474a4", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted"], "repository.name": "ksql", "repository.owner.login": "VinGarcia", "repository.owner.type": "User", "enterprise.slug": "", "action": "queued", "error": "cache contained <nil>, which is not an Object"}
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2047
net/http.(*ServeMux).ServeHTTP
        /usr/local/go/src/net/http/server.go:2425
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2879
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1930
```

Then I copied the logs, deleted both resources and created them again this time with the namespace field set, and it worked normally, the logs showed normal event processing messages:

```logs
2022-05-17T23:30:46Z    DEBUG   controllers.webhookbasedautoscaler      Found 1 HRAs by key     {"key": "VinGarcia/ksql"}
2022-05-17T23:30:46Z    INFO    controllers.webhookbasedautoscaler      job scale up target is repository-wide runners  {"event": "workflow_job", "hookID": "353945701", "delivery": "60751ad0-d639-11ec-823d-b0f433c135ab", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted"], "repository.name": "ksql", "repository.owner.login": "VinGarcia", "repository.owner.type": "User", "enterprise.slug": "", "action": "queued", "repository": "ksql"}
2022-05-17T23:30:46Z    DEBUG   controllers.webhookbasedautoscaler      Patching hra example-runners for capacityReservations update    {"before": 0, "expired": 0, "amount": 1, "after": 1}
2022-05-17T23:30:46Z    INFO    controllers.webhookbasedautoscaler      scaled example-runners by 1
```

Reply:
@VinGarcia Thanks for your detailed report 🙏 

The error you encountered, `cache contained <nil>, which is not an Object`, is very interesting. It reminds me of the same looking error that we concluded to occur due to possibly a hidden bug in `controller-runtime` which we use as the framework for K8s controllers(like ARC).

To see the nature of the problem, would you mind recreating the runnerdeployments "without namespace" once again, and restart ARC, by removing the controller-manager pod and webhookbasedautoscaler 's pod manually? (THose pods are managed by K8s Deployments so K8s would automatically bring up another pods as replacements, hence restart)

Reply:
Oh sure, so, I just made this experiment, and it worked this time (although it took some time to actually start working, at first it was reporting `Found 0 HRAs by key     {"key": "VinGarcia/ksql"}`, and only after some time it started detecting the HRA and scaling normally.

It was a good thing that I waited for a while before writing down the results.

Ok, so I think we can remove the namespace too.